druid/src/widget/env_scope.rs

```diff
@@ -48,61 +54,109 @@ impl<T, W: Widget<T>> EnvScope<T, W> {
     /// # }
     /// ```
     ///
-    /// [`WidgetExt::env_scope`]: ../trait.WidgetExt.html#method.env_scope
-    pub fn new(f: impl Fn(&mut Env, &T) + 'static, child: W) -> EnvScope<T, W> {
+    /// [`WidgetExt::env_scope`]: crate::WidgetExt::env_scope
+    pub fn new(overrides: Env, child: W) -> EnvScope<T, W> {
         EnvScope {
-            f: Box::new(f),
             child: WidgetPod::new(child),
+            current_child_env: None,
+            prev_super_env: None,
+            overrides,
+            modify_env: None,
+            should_modify_env_now: None,
         }
     }
+
+    fn child_env(&mut self, super_env: &Env) {
+        let super_same = self
+            .prev_super_env
+            .as_ref()
+            .map(|old| old.same(super_env))
+            .unwrap_or(false);
+        if !super_same {
+            self.current_child_env = Some(super_env.with_overrides(&self.overrides));
+            self.prev_super_env = Some(super_env.to_owned());
+        }
+    }
+
+    /// This function takes a closure that allows you to modify the Env
+    /// based on the provided data.
+    ///
+    /// The first closure gives you access to app data and a mutable `Env`
+    /// which will be passed to `EnvScope`'s children.
+    ///
+    /// The second closure will determine if the first closure is called
+    /// when there are updates to the provided data.
+    pub fn modify_env(
```

Comment:
Maybe this should be an alternate constructor method? Like: Maybe we have,

```rust
fn new(overrides: Env) -> Self {}
fn dynamic(
    modify_env: impl Fn(&T, &mut Env) + 'static,
    should_modify_env: impl Fn(&T, &T, &Env) -> bool + 'static,
) -> Self {}
```

?

Reply:
Yes this sounds good. It depends on the above conversation. If this becomes dynamic or static then I'll create this constructor.